crates/astria-core/src/sequencerblock/v1/block/mod.rs

```diff
@@ -1165,122 +1162,424 @@ impl ExtendedCommitInfoError {
-    rollup_transactions_root: [u8; 32],
-    rollup_transactions_proof: merkle::Proof,
-    rollup_ids_root: [u8; 32],
-    rollup_ids_proof: merkle::Proof,
+#[derive(Debug, thiserror::Error)]
+#[error(transparent)]
+pub struct DataItemError(DataItemErrorKind);
+
+impl DataItemError {
+    fn encode(source: io::Error, variant: &'static str) -> Self {
+        Self(DataItemErrorKind::Encode {
+            source,
+            variant,
+        })
+    }
+
+    fn decode(source: io::Error, index: usize) -> Self {
+        Self(DataItemErrorKind::Decode {
+            source,
+            index,
+        })
+    }
+
+    fn mismatch(index: usize, expected: &'static str, actual: &'static str) -> Self {
+        Self(DataItemErrorKind::Mismatch {
+            index,
+            expected,
+            actual,
+        })
+    }
 }
 
-fn take_initial_elements_from_data(
-    data: Vec<Bytes>,
-    with_extended_commit_info: bool,
-) -> Result<InitialDataElements, SequencerBlockError> {
-    let tree = merkle_tree_from_data(&data);
-    let data_root_hash = tree.root();
-    let mut data_iter = data.into_iter();
-
-    let rollup_transactions_root: [u8; 32] = data_iter
-        .next()
-        .ok_or(SequencerBlockError::no_rollup_transactions_root())?
-        .as_ref()
-        .try_into()
-        .map_err(|_| {
-            SequencerBlockError::incorrect_rollup_transactions_root_length(data_iter.len())
-        })?;
-    let rollup_transactions_proof = tree.construct_proof(0).expect(
-        "the leaf must exist in the tree as `rollup_transactions_root` was created from the same \
-         index in `data` used to construct the tree",
-    );
-
-    let rollup_ids_root: [u8; 32] = data_iter
-        .next()
-        .ok_or(SequencerBlockError::no_rollup_ids_root())?
-        .as_ref()
-        .try_into()
-        .map_err(|_| SequencerBlockError::incorrect_rollup_ids_root_length(data_iter.len()))?;
-    let rollup_ids_proof = tree.construct_proof(1).expect(
-        "the leaf must exist in the tree as `rollup_ids_root` was created from the same index in \
-         `data` used to construct the tree",
-    );
-
-    let (extended_commit_info, extended_commit_info_proof) = if with_extended_commit_info {
-        let extended_commit_info = data_iter
+#[derive(Debug, thiserror::Error)]
+enum DataItemErrorKind {
+    #[error("failed to borsh-encode {variant} as data item")]
+    Encode {
+        source: io::Error,
+        variant: &'static str,
+    },
+    #[error("item {index} of cometbft `block.data` could not be borsh-decoded")]
+    Decode { source: io::Error, index: usize },
+    #[error("unexpected data element `{actual}` at index {index}, expected `{expected}`")]
+    Mismatch {
+        index: usize,
+        expected: &'static str,
+        actual: &'static str,
+    },
+}
+
+/// The parsed elements of the `data` field of a `SequencerBlock`, along with values derived from
+/// these.
+#[derive(Debug)]
+pub struct ParsedDataItems {
```

Comment:
nit but how about naming this `ExpandedBlockData` or something like that? i think "parsed" is a bit confusing since there's items that are generated not parsed, and `DataItems` makes me think of only the types that are wrapped in `DataItem` and not the user txs

Reply:
Done in [4a045d3](https://github.com/astriaorg/astria/pull/1917/commits/4a045d36c177676c3af883aed37da3cce61f30b6).